File: admin/src/lib/oauthProviders.js
// OAuth provider registry. Each entry describes the authorize-URL builder
// and the token-exchange call. Add new providers here.
//
// All flows use PKCE (S256) because every modern provider supports it and
// it lets the client_secret be optional for public clients.

import crypto from 'node:crypto';
import { request } from 'undici';
import { ValidationError, UpstreamError } from '@lazirouter-cloud/shared';

const PROVIDERS = {
  // Test-only provider; mocks PKCE flow when MOCK_OAUTH_BASE is set.
  mock: {
    authorizeUrl:  (process.env.MOCK_OAUTH_BASE || 'http://localhost:31995') + '/authorize',
    tokenEndpoint: (process.env.MOCK_OAUTH_BASE || 'http://localhost:31995') + '/token',
    scope: 'mock',
    clientIdEnv: 'MOCK_OAUTH_CLIENT_ID',
    clientSecretEnv: 'MOCK_OAUTH_CLIENT_SECRET',
  },

  // ── Claude Code subscription (claude.ai OAuth) ─────────────────────
  // Uses the public claude-cli client_id. Token exchange is JSON, not
  // form-urlencoded; authorize page expects code=true and a literal
  // claude.ai domain. The resulting access_token can call
  // https://api.anthropic.com/v1/messages directly as Bearer auth.
  claude: {
    authorizeUrl:  'https://claude.ai/oauth/authorize',
    tokenEndpoint: 'https://api.anthropic.com/v1/oauth/token',
    scope: 'org:create_api_key user:profile user:inference',
    // The claude-cli's client_id is public, hardcoded — no env var override
    // needed for it to be "ready".
    clientId: '9d1c250a-e61b-44d9-88ed-5944d1962f5e',
    tokenExchangeFormat: 'json',
    authorizeExtraParams: { code: 'true' },
  },

  // ── ChatGPT/Codex subscription (auth.openai.com OAuth) ─────────────
  // Uses the public codex-cli client_id. Standard PKCE with extras the
  // OpenAI Codex CLI sends. Resulting access_token calls
  // https://chatgpt.com/backend-api/codex/responses (Responses API).
  codex: {
    authorizeUrl:  'https://auth.openai.com/oauth/authorize',
    tokenEndpoint: 'https://auth.openai.com/oauth/token',
    scope: 'openid profile email offline_access',
    clientId: 'app_EMoamEEZ73f0CkXaXp7hrann',
    tokenExchangeFormat: 'form',
    authorizeExtraParams: {
      id_token_add_organizations: 'true',
      codex_cli_simplified_flow: 'true',
      originator: 'codex_cli_rs',
    },
  },

  // Configurable providers (env-gated)
  openai: {
    authorizeUrl:  'https://auth.openai.com/authorize',
    tokenEndpoint: 'https://auth.openai.com/oauth/token',
    scope: 'openid offline_access profile email',
    clientIdEnv: 'OPENAI_OAUTH_CLIENT_ID',
    clientSecretEnv: 'OPENAI_OAUTH_CLIENT_SECRET',
  },
  anthropic: {
    authorizeUrl:  'https://console.anthropic.com/oauth/authorize',
    tokenEndpoint: 'https://console.anthropic.com/v1/oauth/token',
    scope: 'org:create_api_key user:profile user:inference',
    clientIdEnv: 'ANTHROPIC_OAUTH_CLIENT_ID',
    clientSecretEnv: 'ANTHROPIC_OAUTH_CLIENT_SECRET',
  },
  github: {
    authorizeUrl:  'https://github.com/login/oauth/authorize',
    tokenEndpoint: 'https://github.com/login/oauth/access_token',
    scope: 'read:user',
    clientIdEnv: 'GITHUB_OAUTH_CLIENT_ID',
    clientSecretEnv: 'GITHUB_OAUTH_CLIENT_SECRET',
  },
};

export function listProviders() {
  return Object.keys(PROVIDERS);
}

export function getProvider(name) {
  const p = PROVIDERS[name];
  if (!p) throw new ValidationError(`Unsupported OAuth provider: ${name}`,
    { supported: Object.keys(PROVIDERS) });

  // Hardcoded clientId (subscription providers) vs env-gated
  const clientId = p.clientId || (p.clientIdEnv ? process.env[p.clientIdEnv] : null);
  if (!clientId) {
    throw new ValidationError(
      `Provider ${name} is not configured on this server`,
      { hint: `set env ${p.clientIdEnv}` }
    );
  }
  return {
    name,
    authorizeUrl: p.authorizeUrl,
    tokenEndpoint: p.tokenEndpoint,
    scope: p.scope,
    clientId,
    clientSecret: p.clientSecret || (p.clientSecretEnv ? process.env[p.clientSecretEnv] : null),
    tokenExchangeFormat: p.tokenExchangeFormat || 'form',
    authorizeExtraParams: p.authorizeExtraParams || {},
  };
}

/** Build a PKCE pair: code_verifier + code_challenge (S256). */
export function buildPkce() {
  const verifier = crypto.randomBytes(32).toString('base64url');
  const challenge = crypto.createHash('sha256').update(verifier).digest('base64url');
  return { verifier, challenge };
}

/** Construct the provider authorize URL the user should be redirected to. */
export function buildAuthorizeUrl(provider, { state, codeChallenge, redirectUri }) {
  const u = new URL(provider.authorizeUrl);
  u.searchParams.set('client_id', provider.clientId);
  u.searchParams.set('response_type', 'code');
  u.searchParams.set('redirect_uri', redirectUri);
  u.searchParams.set('state', state);
  u.searchParams.set('code_challenge', codeChallenge);
  u.searchParams.set('code_challenge_method', 'S256');
  if (provider.scope) u.searchParams.set('scope', provider.scope);
  for (const [k, v] of Object.entries(provider.authorizeExtraParams || {})) {
    u.searchParams.set(k, String(v));
  }
  return u.toString();
}

/**
 * Exchange the authorization code for access + refresh tokens.
 * Some providers (notably claude.ai) require JSON body; pass
 * provider.tokenExchangeFormat = 'json' to use it.
 */
export async function exchangeCode(provider, { code, codeVerifier, redirectUri, state }) {
  // Claude embeds state in the code as `code#state`; strip it back out so
  // we send only the authorization code (the canonical reference behavior).
  let authCode = code;
  let codeState = state;
  if (authCode.includes('#')) {
    const parts = authCode.split('#');
    authCode = parts[0];
    codeState = parts[1] || state;
  }

  let body, contentType;
  if (provider.tokenExchangeFormat === 'json') {
    const payload = {
      grant_type: 'authorization_code',
      code: authCode,
      state: codeState,
      redirect_uri: redirectUri,
      client_id: provider.clientId,
      code_verifier: codeVerifier,
    };
    if (provider.clientSecret) payload.client_secret = provider.clientSecret;
    body = JSON.stringify(payload);
    contentType = 'application/json';
  } else {
    const params = new URLSearchParams();
    params.set('grant_type', 'authorization_code');
    params.set('code', authCode);
    params.set('redirect_uri', redirectUri);
    params.set('client_id', provider.clientId);
    params.set('code_verifier', codeVerifier);
    if (provider.clientSecret) params.set('client_secret', provider.clientSecret);
    body = params.toString();
    contentType = 'application/x-www-form-urlencoded';
  }

  const res = await request(provider.tokenEndpoint, {
    method: 'POST',
    headers: { 'content-type': contentType, 'accept': 'application/json' },
    body,
  });
  const text = await res.body.text();
  if (res.statusCode >= 400) {
    throw new UpstreamError(`Token exchange failed (${res.statusCode}): ${text.slice(0, 300)}`,
      { status_code: res.statusCode });
  }
  let data;
  try { data = JSON.parse(text); }
  catch { data = Object.fromEntries(new URLSearchParams(text)); }
  return data;
}

File: internal/admin/oauth_providers.go
package admin

import (
	"bytes"
	"context"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"

	"github.com/amaozhao/lazirouter/internal/errs"
)

// providerSpec is the static definition of an OAuth provider. One of
// ClientID or ClientIDEnv must be present for the provider to be usable.
type providerSpec struct {
	AuthorizeURL         string
	TokenEndpoint        string
	Scope                string
	ClientID             string
	ClientIDEnv          string
	ClientSecret         string
	ClientSecretEnv      string
	TokenExchangeFormat  string // "json" or "form" (default form)
	AuthorizeExtraParams map[string]string
}

// resolvedProvider is providerSpec with env lookups applied.
type resolvedProvider struct {
	Name                 string
	AuthorizeURL         string
	TokenEndpoint        string
	Scope                string
	ClientID             string
	ClientSecret         string
	TokenExchangeFormat  string
	AuthorizeExtraParams map[string]string
}

func providerSpecs() map[string]providerSpec {
	mockBase := os.Getenv("MOCK_OAUTH_BASE")
	if mockBase == "" {
		mockBase = "http://localhost:31995"
	}
	return map[string]providerSpec{
		"mock": {
			AuthorizeURL:    mockBase + "/authorize",
			TokenEndpoint:   mockBase + "/token",
			Scope:           "mock",
			ClientIDEnv:     "MOCK_OAUTH_CLIENT_ID",
			ClientSecretEnv: "MOCK_OAUTH_CLIENT_SECRET",
		},
		"claude": {
			AuthorizeURL:        "https://claude.ai/oauth/authorize",
			TokenEndpoint:       "https://api.anthropic.com/v1/oauth/token",
			Scope:               "org:create_api_key user:profile user:inference",
			ClientID:            "9d1c250a-e61b-44d9-88ed-5944d1962f5e",
			TokenExchangeFormat: "json",
			AuthorizeExtraParams: map[string]string{
				"code": "true",
			},
		},
		"codex": {
			AuthorizeURL:        "https://auth.openai.com/oauth/authorize",
			TokenEndpoint:       "https://auth.openai.com/oauth/token",
			Scope:               "openid profile email offline_access",
			ClientID:            "app_EMoamEEZ73f0CkXaXp7hrann",
			TokenExchangeFormat: "form",
			AuthorizeExtraParams: map[string]string{
				"id_token_add_organizations": "true",
				"codex_cli_simplified_flow":  "true",
				"originator":                 "codex_cli_rs",
			},
		},
		"openai": {
			AuthorizeURL:    "https://auth.openai.com/authorize",
			TokenEndpoint:   "https://auth.openai.com/oauth/token",
			Scope:           "openid offline_access profile email",
			ClientIDEnv:     "OPENAI_OAUTH_CLIENT_ID",
			ClientSecretEnv: "OPENAI_OAUTH_CLIENT_SECRET",
		},
		"anthropic": {
			AuthorizeURL:    "https://console.anthropic.com/oauth/authorize",
			TokenEndpoint:   "https://console.anthropic.com/v1/oauth/token",
			Scope:           "org:create_api_key user:profile user:inference",
			ClientIDEnv:     "ANTHROPIC_OAUTH_CLIENT_ID",
			ClientSecretEnv: "ANTHROPIC_OAUTH_CLIENT_SECRET",
		},
		"github": {
			AuthorizeURL:    "https://github.com/login/oauth/authorize",
			TokenEndpoint:   "https://github.com/login/oauth/access_token",
			Scope:           "read:user",
			ClientIDEnv:     "GITHUB_OAUTH_CLIENT_ID",
			ClientSecretEnv: "GITHUB_OAUTH_CLIENT_SECRET",
		},
	}
}

func listProviders() []string {
	specs := providerSpecs()
	out := make([]string, 0, len(specs))
	for name := range specs {
		out = append(out, name)
	}
	return out
}

// resolveProvider applies env lookups and reports if the provider is ready.
func resolveProvider(name string) (*resolvedProvider, error) {
	specs := providerSpecs()
	p, ok := specs[name]
	if !ok {
		names := listProviders()
		return nil, errs.Validation(fmt.Sprintf("Unsupported OAuth provider: %s", name),
			map[string]any{"supported": names})
	}
	cid := p.ClientID
	if cid == "" && p.ClientIDEnv != "" {
		cid = os.Getenv(p.ClientIDEnv)
	}
	if cid == "" {
		return nil, errs.Validation(fmt.Sprintf("Provider %s is not configured on this server", name),
			map[string]any{"hint": "set env " + p.ClientIDEnv})
	}
	cs := p.ClientSecret
	if cs == "" && p.ClientSecretEnv != "" {
		cs = os.Getenv(p.ClientSecretEnv)
	}
	fmtt := p.TokenExchangeFormat
	if fmtt == "" {
		fmtt = "form"
	}
	return &resolvedProvider{
		Name:                 name,
		AuthorizeURL:         p.AuthorizeURL,
		TokenEndpoint:        p.TokenEndpoint,
		Scope:                p.Scope,
		ClientID:             cid,
		ClientSecret:         cs,
		TokenExchangeFormat:  fmtt,
		AuthorizeExtraParams: p.AuthorizeExtraParams,
	}, nil
}

// buildPKCE creates a verifier + S256 challenge (base64url, no padding).
func buildPKCE() (verifier, challenge string, err error) {
	b := make([]byte, 32)
	if _, err = rand.Read(b); err != nil {
		return "", "", err
	}
	verifier = base64.RawURLEncoding.EncodeToString(b)
	sum := sha256.Sum256([]byte(verifier))
	challenge = base64.RawURLEncoding.EncodeToString(sum[:])
	return
}

func randomState() (string, error) {
	b := make([]byte, 24)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(b), nil
}

func buildAuthorizeURL(p *resolvedProvider, state, challenge, redirectURI string) (string, error) {
	u, err := url.Parse(p.AuthorizeURL)
	if err != nil {
		return "", err
	}
	q := u.Query()
	q.Set("client_id", p.ClientID)
	q.Set("response_type", "code")
	q.Set("redirect_uri", redirectURI)
	q.Set("state", state)
	q.Set("code_challenge", challenge)
	q.Set("code_challenge_method", "S256")
	if p.Scope != "" {
		q.Set("scope", p.Scope)
	}
	for k, v := range p.AuthorizeExtraParams {
		q.Set(k, v)
	}
	u.RawQuery = q.Encode()
	return u.String(), nil
}

// exchangeCode posts code + verifier to the token endpoint and returns the
// parsed response map. Tolerates form-encoded responses (GitHub).
func exchangeCode(ctx context.Context, p *resolvedProvider, code, verifier, redirectURI, state string) (map[string]any, error) {
	// claude.ai mangles state into the code as `code#state`; strip it.
	authCode := code
	if i := strings.Index(authCode, "#"); i >= 0 {
		authCode = code[:i]
	}

	var body []byte
	var ctype string
	if p.TokenExchangeFormat == "json" {
		payload := map[string]any{
			"grant_type":    "authorization_code",
			"code":          authCode,
			"state":         state,
			"redirect_uri":  redirectURI,
			"client_id":     p.ClientID,
			"code_verifier": verifier,
		}
		if p.ClientSecret != "" {
			payload["client_secret"] = p.ClientSecret
		}
		body, _ = json.Marshal(payload)
		ctype = "application/json"
	} else {
		v := url.Values{}
		v.Set("grant_type", "authorization_code")
		v.Set("code", authCode)
		v.Set("redirect_uri", redirectURI)
		v.Set("client_id", p.ClientID)
		v.Set("code_verifier", verifier)
		if p.ClientSecret != "" {
			v.Set("client_secret", p.ClientSecret)
		}
		body = []byte(v.Encode())
		ctype = "application/x-www-form-urlencoded"
	}

	req, _ := http.NewRequestWithContext(ctx, http.MethodPost, p.TokenEndpoint, bytes.NewReader(body))
	req.Header.Set("content-type", ctype)
	req.Header.Set("accept", "application/json")

	cli := &http.Client{Timeout: 20 * time.Second}
	res, err := cli.Do(req)
	if err != nil {
		return nil, errs.Upstream("token exchange transport failed: " + err.Error())
	}
	defer res.Body.Close()
	buf, _ := io.ReadAll(res.Body)
	if res.StatusCode >= 400 {
		snippet := string(buf)
		if len(snippet) > 300 {
			snippet = snippet[:300]
		}
		return nil, errs.Upstream(fmt.Sprintf("Token exchange failed (%d): %s", res.StatusCode, snippet),
			map[string]any{"status_code": res.StatusCode})
	}
	var data map[string]any
	if err := json.Unmarshal(buf, &data); err == nil {
		return data, nil
	}
	// Fallback: form-encoded (GitHub).
	v, ferr := url.ParseQuery(string(buf))
	if ferr != nil {
		return nil, errs.Upstream("token response is neither JSON nor form-encoded",
			map[string]any{"body": string(buf)})
	}
	out := make(map[string]any, len(v))
	for k := range v {
		out[k] = v.Get(k)
	}
	return out, nil
}
